File: src/document/loader.py
"""Document loader"""

import re
import logging
from pathlib import Path
from typing import Dict, Iterator, List, Optional
from dataclasses import dataclass, field

logger = logging.getLogger(__name__)


def clean_html_from_markdown(content: str) -> str:
	"""
	Clean HTML tags and CSS styles from Markdown

	Keep:
	- Code blocks (```...```)
	- Image tags ![...](...) and <img>
	- Link tags [...](...) and <a>

	Remove:
	- <span style="...">text</span> → text
	- <br> → newline
	- Other meaningless HTML tags
	"""
	# Protect code blocks
	code_blocks = []

	def save_code_block(match):
		code_blocks.append(match.group(0))
		return f"__CODE_BLOCK_{len(code_blocks) - 1}__"

	content = re.sub(r"```[\s\S]*?```", save_code_block, content)

	# Remove <span> tags but keep content
	content = re.sub(r"<span[^>]*>([^<]*)</span>", r"\1", content)

	# Handle nested spans (multiple passes)
	for _ in range(5):
		prev = content
		content = re.sub(r"<span[^>]*>([^<]*)</span>", r"\1", content)
		if content == prev:
			break

	# Remove remaining empty or complex spans
	content = re.sub(r"<span[^>]*>\s*</span>", "", content)
	content = re.sub(r"<span[^>]*>", "", content)
	content = re.sub(r"</span>", "", content)

	# <br> → newline
	content = re.sub(r"<br\s*/?>", "\n", content)

	# Remove Word export attributes
	content = re.sub(r'\s*data-ccp-props="[^"]*"', "", content)

	# Clean style attributes
	content = re.sub(r'\s*style="[^"]*"', "", content)
	content = re.sub(r'\s*class="[^"]*"', "", content)

	# Clean excess whitespace
	content = re.sub(r"\n{3,}", "\n\n", content)
	content = re.sub(r" {2,}", " ", content)

	# Restore code blocks
	for i, block in enumerate(code_blocks):
		content = content.replace(f"__CODE_BLOCK_{i}__", block)

	return content.strip()


@dataclass
class Document:
	"""Document object"""

	id: str
	content: str
	metadata: Dict = field(default_factory=dict)


class DocumentLoader:
	"""
	Load Markdown documents from directory, auto-extract directory structure as metadata

	Directory structure example:
	raw_data/
	├── apis/          # category="api"
	├── docs/          # category="doc"
	└── demos/         # category="demo"

	Extracted metadata:
	- category: api/doc/demo
	- path_hierarchy: full path hierarchy list
	- file_name: filename (without extension)
	"""

	CATEGORY_MAP = {
		"apis": "api",
		"docs": "doc",
		"demos": "demo",
	}

	def __init__(self, base_dir: str | Path):
		self.base_dir = Path(base_dir)
		if not self.base_dir.exists():
			raise ValueError(f"Directory not found: {base_dir}")

	def _extract_metadata(self, file_path: Path) -> Dict:
		"""Extract metadata from file path"""
		rel_path = file_path.relative_to(self.base_dir)
		parts = list(rel_path.parts)

		metadata = {
			"file_path": str(file_path),
			"relative_path": str(rel_path),
			"file_name": file_path.stem,
			"path_hierarchy": parts[:-1],
		}

		if parts:
			top_dir = parts[0].lower()
			metadata["category"] = self.CATEGORY_MAP.get(top_dir, top_dir)

		return metadata

	def load_file(self, file_path: Path) -> Optional[Document]:
		"""Load single file"""
		try:
			content = file_path.read_text(encoding="utf-8")
			if not content.strip():
				return None

			content = clean_html_from_markdown(content)
			metadata = self._extract_metadata(file_path)

			# Generate document ID
			doc_id = metadata["relative_path"].replace("\\", "/").replace("/", "_")
			for ext in [".md", ".java"]:
				doc_id = doc_id.replace(ext, "")

			return Document(id=doc_id, content=content, metadata=metadata)
		except Exception as e:
			logger.warning(f"Failed to load file {file_path}: {e}")
			return None

	def load_directory(
		self,
		subdirs: Optional[List[str]] = None,
		extensions: List[str] | None = None,
	) -> Iterator[Document]:
		"""Load documents from directory"""
		if extensions is None:
			extensions = [".md", ".java"]

		search_dirs = []
		if subdirs:
			for subdir in subdirs:
				d = self.base_dir / subdir
				if d.exists():
					search_dirs.append(d)
		else:
			search_dirs = [self.base_dir]

		total_count = 0
		for search_dir in search_dirs:
			for ext in extensions:
				for file_path in search_dir.rglob(f"*{ext}"):
					if not file_path.is_file():
						continue
					doc = self.load_file(file_path)
					if doc:
						total_count += 1
						yield doc

		logger.info(f"Load complete, {total_count} documents total")

	def load_all(self, subdirs: Optional[List[str]] = None) -> List[Document]:
		"""Load all documents to list"""
		return list(self.load_directory(subdirs))

File: src/document/loader.ts
/**
 * Document loader for GC-DOC-MCP v2
 * Loads Markdown documents from directory, extracts directory structure as metadata
 */

import { promises as fs } from 'fs';
import * as path from 'path';
import { Document, DocumentCategory } from './types.js';
import { createDefaultLogger } from '../shared/logger.js';

const logger = createDefaultLogger('document:loader');

const CATEGORY_MAP: Record<string, DocumentCategory> = {
	apis: 'api',
	docs: 'doc',
	demos: 'demo',
};

/**
 * Clean HTML tags and CSS styles from Markdown
 * Keep: code blocks, images, links
 * Remove: span tags, br, style attributes, Word export attributes
 */
export function cleanHtmlFromMarkdown(content: string): string {
	// Protect code blocks
	const codeBlocks: string[] = [];

	const saveCodeBlock = (match: string): string => {
		codeBlocks.push(match);
		return `__CODE_BLOCK_${codeBlocks.length - 1}__`;
	};

	let cleaned = content.replace(/```[\s\S]*?```/g, saveCodeBlock);

	// Handle nested spans (multiple passes)
	const MAX_NESTED_SPAN_PASSES = 5;
	for (let i = 0; i < MAX_NESTED_SPAN_PASSES; i++) {
		const prev = cleaned;
		cleaned = cleaned.replace(/<span[^>]*>([^<]*)<\/span>/g, '$1');
		if (cleaned === prev) break;
	}

	// Remove remaining empty or complex spans
	cleaned = cleaned.replace(/<span[^>]*>\s*<\/span>/g, '');
	cleaned = cleaned.replace(/<span[^>]*>/g, '');
	cleaned = cleaned.replace(/<\/span>/g, '');

	// <br> → newline
	cleaned = cleaned.replace(/<br\s*\/?>/g, '\n');

	// Remove Word export attributes
	cleaned = cleaned.replace(/\s*data-ccp-props="[^"]*"/g, '');

	// Clean style attributes
	cleaned = cleaned.replace(/\s*style="[^"]*"/g, '');
	cleaned = cleaned.replace(/\s*class="[^"]*"/g, '');

	// Clean excess whitespace
	cleaned = cleaned.replace(/\n{3,}/g, '\n\n');
	cleaned = cleaned.replace(/ {2,}/g, ' ');

	// Restore code blocks
	for (let i = 0; i < codeBlocks.length; i++) {
		cleaned = cleaned.replace(`__CODE_BLOCK_${i}__`, codeBlocks[i]);
	}

	return cleaned.trim();
}

export class DocumentLoader {
	private readonly baseDir: string;
	private readonly categoryMap: Record<string, DocumentCategory>;

	constructor(baseDir: string) {
		this.baseDir = baseDir;
		this.categoryMap = CATEGORY_MAP;
	}

	private extractMetadata(filePath: string): Document['metadata'] {
		const relPath = path.relative(this.baseDir, filePath);
		const parts = relPath.split(path.sep).filter(Boolean);

		const metadata: Document['metadata'] = {
			file_path: filePath,
			relative_path: relPath,
			file_name: path.basename(filePath, path.extname(filePath)),
			path_hierarchy: parts.slice(0, -1),
		};

		if (parts.length > 0) {
			const topDir = parts[0].toLowerCase();
			metadata.category = this.categoryMap[topDir] ?? (topDir as DocumentCategory);
		}

		return metadata;
	}

	private async loadFile(filePath: string): Promise<Document | null> {
		try {
			let content = await fs.readFile(filePath, 'utf-8');

			if (!content.trim()) {
				return null;
			}

			content = cleanHtmlFromMarkdown(content);
			const metadata = this.extractMetadata(filePath);

			// Generate document ID from relative path
			let docId = metadata.relative_path.replace(/\\/g, '/').replace(/\//g, '_');
			docId = docId.replace(/\.[^.]+$/, '');

			return {
				id: docId,
				content,
				metadata,
			};
		} catch (e) {
			logger.warn(`Failed to load file ${filePath}: ${e}`);
			return null;
		}
	}

	public async *loadDirectory(
		subdirs?: string[],
		extensions: string[] = ['.md', '.java'],
	): AsyncGenerator<Document> {
		const searchDirs: string[] = [];

		if (subdirs) {
			for (const subdir of subdirs) {
				const dir = path.join(this.baseDir, subdir);
				try {
					const stat = await fs.stat(dir);
					if (stat.isDirectory()) {
						searchDirs.push(dir);
					}
				} catch {
					// Directory doesn't exist, skip
				}
			}
		} else {
			searchDirs.push(this.baseDir);
		}

		for (const searchDir of searchDirs) {
			for (const ext of extensions) {
				const files = await this.findFiles(searchDir, ext);
				for (const filePath of files) {
					const doc = await this.loadFile(filePath);
					if (doc) {
						yield doc;
					}
				}
			}
		}
	}

	private async findFiles(dir: string, ext: string): Promise<string[]> {
		const files: string[] = [];
		const entries = await fs.readdir(dir, { withFileTypes: true });

		for (const entry of entries) {
			const fullPath = path.join(dir, entry.name);

			if (entry.isDirectory()) {
				const subFiles = await this.findFiles(fullPath, ext);
				files.push(...subFiles);
			} else if (entry.isFile() && entry.name.endsWith(ext)) {
				files.push(fullPath);
			}
		}

		return files;
	}

	public async loadAll(subdirs?: string[]): Promise<Document[]> {
		const docs: Document[] = [];
		for await (const doc of this.loadDirectory(subdirs)) {
			docs.push(doc);
		}
		return docs;
	}
}
